### ISSUE TITLE
Boundary labels, text and line orientation flips

### ISSUE THREAD

**Issue by sensescape** [tangrams/tangram#471]
Added boundary labels on branches for places refactor on all the house styles.

Text orientation looks like this:
<img width="491" alt="screen shot 2016-12-15 at 1 29 28 pm" src="https://cloud.githubusercontent.com/assets/466585/21242799/e2682826-c2ca-11e6-9f0f-5ca8bec90f12.png">
here, Utah is on both sides

Here's another one:
<img width="755" alt="screen shot 2016-12-15 at 1 33 08 pm" src="https://cloud.githubusercontent.com/assets/466585/21242844/1a2f8010-c2cb-11e6-9056-559212ac1511.png">

**Comment by bcamper** [comment]
@sensescape can you send a link to the YAML you're using for this?

@nvkelso may be a data issue, since Tangram won't know which way the line is "supposed" to go - or it could just be that Tangram is flipping the text orientation to keep it upright... in which case some additional styling and/or data hints would be needed so that it knows what it can move where.

**Comment by sensescape** [comment]
@bcamper @nvkelso 
another example:
<img width="408" alt="screen shot 2016-12-15 at 1 52 34 pm" src="https://cloud.githubusercontent.com/assets/466585/21243467/d95ce53e-c2cd-11e6-9b5c-2870b0fcb609.png">

https://github.com/tangrams/refill-style/blob/sensescape/places-refactor/refill-style.yaml#L4014-L4040

https://github.com/tangrams/refill-style/blob/sensescape/places-refactor/refill-style.yaml#L4068-L4095

**Comment by meetar** [comment]
iirc at one point we discussed having a label option for "inside" vs
"outside" …

**Comment by bcamper** [comment]
@sensescape a couple related notes on syntax issues in the snippets you linked (these should be covered in the Tangram docs, we can clarify if not):

`text_wrap: 100px`
This parameter is the **number of characters** to wrap on, not the number of pixels. 

`offset: 10px`
This parameter requires **two values**, for x and y: `offset: [0px, 10px]` (offsetting a line label in the x direction probably isn't super useful, but it's technically possible and provides consistency on the `offset` parameter for point labels).

**Comment by sensescape** [comment]
@bcamper 
ok, thanks!
will remove px and rework offset

**Comment by nvkelso** [comment]
@bcamper We're still seeing inverted text (where left labels right, right labels left). Holding off on releasing anything related to this in Mapzen house styles until there's a fix (not urgent).

**Comment by bcamper** [comment]
@nvkelso explicit left and right boundary labels are not something Tangram currently supports, though we have discussed adding support.

When we added the `offset` parameter for text, we discussed this and decided to interpret it as an offset based on the **text baseline**, not based on the **line geometry**. I found an old Trello (!) card with discussion on this, and conclusion that the current behavior was desirable for standard cases, with boundaries being additional functionality we would add in the future (at the time we didn't have this data in tiles anyway). The Trello card is private but [here](https://trello.com/c/FnQaZnuI/624-line-label-offsets) for reference. Here are two old screenshots from that discussion that illustrate the issue:

Text is always placed such that it sits "on top of" the line (text bottom baseline is aligned with top of line geometry), regardless of the natural direction of the line geometry:
![text-above-line](https://cloud.githubusercontent.com/assets/16733/21429738/2100427c-c82e-11e6-8d0f-1cef1dd32109.png)

Here is what we would need for boundary labels, where the text is placed above or below according to the direction of the line geometry itself (this of course assumes that the line geometries have the proper direction in tiles for admin boundaries to correctly fall on the left and right):
![text-angle-adjust](https://cloud.githubusercontent.com/assets/16733/21429737/20fdf8dc-c82e-11e6-8622-3d8919782adb.png)

Now that we have the data available, we should revisit the feature, and we can use this issue for that. This is something @dmvaldman may be able to look at once his curved label work is merged. In the meantime, we should probably avoid labels with a line break (maybe set a high `text_wrap`?).

We'll have to discuss syntax, perhaps something where a single `text` draw group generates multiple label placement, with separate `text_source` specified for left and right sides. If we wanted to avoid introducing another parameter (which would need precedence rules to disambiguate), we could possibly extend `text_source` to accept an object, like this:

```
draw:
  text:
    text_source:
      left: 'name:left'
      right: 'name:right'
    ...
```

cc @matteblair

**Comment by dmvaldman** [comment]
Thanks for bringing this up. I've been assuming the line orientation was in a sense "irrelevant", but I can keep track of the orientation and make sure the offsets are consistent throughout the line geometry.

**Comment by nvkelso** [comment]
I dig:

draw:
  text:
    text_source:
      left: 'name:left'
      right: ‘name:right'


We enforce the orientation of the boundary line to be consistent on
the server as we calculate the left and right names.

**Comment by dmvaldman** [comment]
Starting to work on this. Not quite there yet though.

<img width="543" alt="screen shot 2017-02-23 at 5 58 41 pm" src="https://cloud.githubusercontent.com/assets/796394/23283026/dd653770-f9f1-11e6-9ab9-6882ef81f7b4.png">

**Comment by bcamper** [comment]
Excite

**Comment by nvkelso** [comment]
These related vector tile issues would be nice to fix so we could abbreviate the region or country names when zoomed out (e.g. `NY` and `NJ` versus `New York` and `New Jersey` – or `N.Y.` and `N.J.`):

- https://github.com/tilezen/vector-datasource/issues/1102
- https://github.com/tilezen/vector-datasource/issues/1094

**Comment by nvkelso** [comment]
Is there a sister issue for Tangram ES?

(And would styling this in Tangram JS break anything in ES until such support lands in ES?)

**Comment by bcamper** [comment]
Released in [v0.12.0](https://github.com/tangrams/tangram/releases/tag/v0.12.0)
